library/psa_crypto_core.h

```diff
@@ -158,6 +158,46 @@ typedef enum {
 #define PSA_KEY_SLOT_MUTEX_LOCKFUNC_RETURN(op)  do { } while (0)
 #endif
 
+/** If threading is enabled: perform a lock or unlock operation on the
+ * key slot mutex.
+ * Call with parameter PSA_MUTEX_LOCK to perform a lock operation.
+ * Call with parameter PSA_MUTEX_UNLOCK to perform an unlock operation.
+ * This will goto the exit label if the operation fails,
+ * setting status to PSA_SERVICE_FAILURE if status was PSA_SUCCESS.
+ * If threading is not enabled, do nothing.
+ *
+ * Assumptions:
+ *  psa_status_t status exists.
+ *  Label exit: exists.
+ *  op is PSA_MUTEX_LOCK or PSA_MUTEX_UNLOCK.
+ */
+#if defined(MBEDTLS_THREADING_C)
+#define PSA_KEY_SLOT_MUTEX_LOCKFUNC_GOTO_EXIT(op)                    \
+    do                                                               \
+    {                                                                \
+        if (op == PSA_MUTEX_LOCK) {                                  \
+            if (mbedtls_mutex_lock(                                  \
+                    &mbedtls_threading_key_slot_mutex) != 0) {       \
+                if (status == PSA_SUCCESS) {                         \
+                    status = PSA_ERROR_SERVICE_FAILURE;              \
+                }                                                    \
+                goto exit;                                           \
+            }                                                        \
+        }                                                            \
+        if (op == PSA_MUTEX_UNLOCK) {                                \
+            if (mbedtls_mutex_unlock(                                \
+                    &mbedtls_threading_key_slot_mutex) != 0) {       \
+                if (status == PSA_SUCCESS) {                         \
+                    status = PSA_ERROR_SERVICE_FAILURE;              \
+                }                                                    \
+                goto exit;                                           \
+            }                                                        \
+        }                                                            \
+    } while (0);
+#else
+#define PSA_KEY_SLOT_MUTEX_LOCKFUNC_GOTO_EXIT(op)   do { } while (0)
+#endif
+
```

Comment:
I think we should go for something simpler, like
```C
#define PSA_THREADING_CHK(f)       \
    do                           \
    {                            \
        if ((ret = (f)) != 0) { \
            if (status == PSA_SUCCESS) {                         \
                status = PSA_ERROR_SERVICE_FAILURE;              \
                }   
            goto cleanup;        \
            }
    } while (0)
```

I can see that
```C
PSA_THREADING_CHK(PSA_MUTEX_LOCK);
```
is more compact than
```C
PSA_THREADING_CHK(mbedtls_mutex_unlock(
                  &mbedtls_threading_key_slot_mutex));
```
But I find it easier to read. It is not much longer and it has one less macro I have to interpret/unwrap in my head. It is usually easier to write code than to read it, so I think the trade-off is worth it.

Reply:
Ah, I missed that we would need that. It is a bit more code, but I agree that it has an upside and it is still easier to read.

Reply:
So for the names would you be happy with `PSA_THREADING_CHK_RET`/`PSA_THREADING_CHK_GOTO_EXIT`?

Reply:
@paul-elliott-arm are you happy for me to make this change?

Reply:
Yep, go ahead.